engine/core/src/main/java/org/datacleaner/util/ReflectionUtils.java

```diff
@@ -331,54 +331,80 @@ public static boolean isWildcard(Type type) {
         throw new UnsupportedOperationException("Parameter type not supported: " + someType);
     }
 
-    public static int getHierarchyDistance(Class<?> subtype, Class<?> supertype) {
+    public static int getHierarchyDistance(Class<?> subtype, Class<?> supertype) throws IllegalArgumentException {
         assert subtype != null;
         assert supertype != null;
 
-        if (subtype == supertype) {
-            return 0;
-        }
-
         if (!ReflectionUtils.is(subtype, supertype)) {
             throw new IllegalArgumentException("Not a valid subtype of " + supertype.getName() + ": "
                     + subtype.getName());
         }
 
-        Class<?> subSuperclass = subtype.getSuperclass();
-        if (subSuperclass != Object.class) {
-            return 1 + getHierarchyDistance(subSuperclass, supertype);
-        }
-
         if (supertype.isInterface()) {
             return getInterfaceHierarchyDistance(subtype, supertype);
         } else {
-            return 1 + getHierarchyDistance(subSuperclass, supertype);
+            return getClassHierarchyDistance(subtype, supertype);
         }
     }
 
-    private static int getInterfaceHierarchyDistance(Class<?> subtype, Class<?> supertype) {
+    private static int getClassHierarchyDistance(final Class<?> subtype, final Class<?> supertype) {
         if (subtype == supertype) {
             return 0;
         }
+        if (subtype == Object.class) {
+            throw new IllegalArgumentException("Finished traversing hierarchy and found Object.class");
```

Comment:
I might be wrong, but this seems like a bad use of exceptions to me; judging by the code below (wrapping them in empty catch blocks), there is nothing exceptional about this, it's just a "not found".